### PR TITLE
[Quantization] Refactoring for 16-bit quantization

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
- Add infrastructure (utility functions) for 16-bit quantization.
- Remove **IntLookupTable** node/instruction definition constraint to have same input/output type (in principle a backend could choose to have an implementation for int8 input and int16 output).
- When replacing quantized Tanh/Sigmoid with Lookup tables make sure the quantization schema is honored when inserting RescaleQuantized for the input/output in Quantization.cpp.
- Generalize the Graph.cpp primitives for creating lookup tables (especially for Sigmoid and Tanh which had some hard-coded assumptions about the input/output quantization parameters which might not have been evident)
- Some generalizations of the quantization primitives from LLVM/CPU libjit.
- Add quantization constraint that input/output parameters must match for additional nodes: ResizeBilinear, ResizeNearest (nodes which reorder data or nodes for which the output range is included in the input range). This is similar to TensorFlowLite (details [here](https://www.tensorflow.org/lite/performance/quantization_spec))
- Remove some hard-coded assumptions from other places about the quantized precision type (GraphOptimizer.cpp, Lower.cpp)

Note: This PR is part from a larger internal PR which is split in multiple pieces for size considerations.

Co-authored by Guido Roncarolo (guido.roncarolo@nxp.com).

**Test Plan**
Work in progress.

### CODE REVIEW

File: include/glow/Base/Type.h [hunk 1]
```diff
@@ -472,6 +472,14 @@ inline ElemKind getScaleOffsetElemKindFromFused(ElemKind e) {
...
                                                ElemKind elementType);
 
+/// \returns the number of values associated to a quantized type given
+/// \p elementType (e.g. 256 for Int8QTy).
+size_t getQuantizedValueCount(ElemKind elementType);
+
+/// \returns the floating point value step associated to a quantized type given
+/// \p scale, \p offset, and \p elementType.
+float getQuantizedValueStep(float scale, int32_t offset, ElemKind elementType);
+
```
Comment (jfix71) [inline include/glow/Base/Type.h:482]: nit: These aren't used aside from inside Type, right? Why not just add them to `Type.{cpp,h}`

File: lib/Importer/ONNXModelLoader.cpp [hunk 2]
```diff
@@ -4099,16 +4099,32 @@ Error ONNXModelLoader::loadIntLookupTable(const ONNX_NAMESPACE::NodeProto &op,
...
+    ASSIGN_VALUE_OR_RETURN_ERR(shape, getShape<dim_t>(dict["shape"]));
+
+    auto outTy = mod_.uniqueType(in.getElementType(), shape);
+    Node *N = G_->createIntLookupTable<int16_t>(loadOperatorName(op), in,
+                                                values, outTy);
+
+    RETURN_IF_ERR(addNodeAsOutput(op, N));
+    return Error::success();
+  }
+  return MAKE_ERR("Invalid option");
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp:4127]: nit: Could make this a little more descriptive of an error, e.g. using `getElementName` to include the type in the error.

**Review by jfix71 — APPROVED** [review]
Looks really great! Just a couple nits.

File: lib/Importer/ONNXModelLoader.cpp (hunk 2 above)
Comment (mciprian13) [reply lib/Importer/ONNXModelLoader.cpp]: Done.

File: include/glow/Base/Type.h (hunk 1 above)
Comment (mciprian13) [reply include/glow/Base/Type.h:482]: Currently these functions are available both outside and inside (as members of) the ```Type``` class (similar to the `getQuantizedValueRange`). Not sure I understand the request ... you mean we should only have them as class members?

**Comment by mciprian13** [comment]
@jfix71 Do you think we could have this imported in your internal build and see whether it breaks something? Thanks!

File: include/glow/Base/Type.h (hunk 1 above)
Comment (jfix71) [reply include/glow/Base/Type.h:482]: Sorry -- yes, I meant right now you forward to them inside `Type:: getQuantizedValueCount` and `Type::getQuantizedValueStep`, but if they're not used outside of the Type class then they could only be class members. Not a huge deal, just seemed unnecessary.

Comment (mciprian13) [reply include/glow/Base/Type.h]: I kept the two functions only inside the Type class. Done!

**Comment by mciprian13** [comment]
@jfix71 I addressed the second item of your review. The failing tests seem totally unrelated.

**Comment by yinghai** [comment]
`glow:BackendCorrectnessTest_NNPI - NNPI/BackendCorrectnessTest.intLookupTableInt16/0`

failed, we probably need to blacklist this one.

**Comment by mciprian13** [comment]
@yinghai I tried black listing the NNPI test but now nothing there are some git issues. Do you know what is the problem?

**Comment by mciprian13** [comment]
@yinghai I added the unit test in the blacklist as instructed. Can you have another look?
Thanks!

**Pull request closed without merging**